src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java

```diff
@@ -824,142 +575,79 @@ private boolean locationIsAnEndPointOfAnIntersectingEdgeOfDifferentLayer(
-        }
-        else if (pointTagSize < AtlasTag.TAGS_FROM_OSM.size())
-        {
-            osmAndAtlasTagCount = pointTagSize;
-        }
-        else
-        {
-            osmAndAtlasTagCount = AtlasTag.TAGS_FROM_OSM.size();
+            logger.error("Edge {} hass less than {} nodes, cannot be sectioned!",
+                    line.getIdentifier(), MINIMUM_NODES_TO_QUALIFY_AS_A_EDGE);
+            this.changes.add(FeatureChange.add(
+                    CompleteLine.shallowFrom(line).withAddedTag(SyntheticInvalidWaySectionTag.KEY,
+                            SyntheticInvalidWaySectionTag.YES.toString()),
+                    this.rawAtlas));
+            return;
         }
 
-        return pointTagSize > osmAndAtlasTagCount;
-    }
+        // Initialize start location
+        PolyLine remainder = null;
 
-    /**
-     * Determines if the given {@link Line} contains more than
-     * {@link #MINIMUM_SHAPE_POINTS_TO_QUALIFY_AS_AREA} shape points. The reason to use 3 is because
-     * OSM forces closed polygons. If we have only 3 nodes (a -> b -> c), then it's not a polygon.
-     * If we have (a -> b -> a), then we have a flat line. At minimum, we are looking for (a -> b ->
-     * c -> a).
-     *
-     * @param line
-     *            The {@link Line} to check
-     * @return {@code true} if given {@link Line} contains more than
-     *         {@link #MINIMUM_SHAPE_POINTS_TO_QUALIFY_AS_AREA} shape points
-     */
-    private boolean qualifiesAsArea(final Line line)
-    {
-        return line.numberOfShapePoints() > MINIMUM_SHAPE_POINTS_TO_QUALIFY_AS_AREA;
-    }
-
-    /**
-     * Sections all lines that are set to become edges in the final atlas.
-     *
-     * @param changeSet
-     *            The {@link WaySectionChangeSet} to rely on for keeping track of updates
-     */
-    private void sectionEdges(final WaySectionChangeSet changeSet)
-    {
-        final Time sectionTime = logTaskStartedAsInfo(EDGE_SECTIONING_TASK, getShardOrAtlasName());
-
-        changeSet.getLinesThatBecomeEdges().forEach(lineIdentifier ->
+        // we'll preprocess rings a bit to make them consistently sectioned
+        if (line.isClosed())
         {
-            final Line line = this.rawAtlas.line(lineIdentifier);
-            List<TemporaryEdge> edges = new ArrayList<>();
-            try
+            if (nodesForEdge.size() == 2)
             {
-                if (line.isClosed())
+                // the node is just the beginning/end of the Edge, and connects to another Edge,
+                // like a cul-de-sac
+                if (nodesForEdge.get(0) == 0
+                        && nodesForEdge.get(1) == line.numberOfShapePoints() - 1
+                        && locationIsAnEndPointOfAnIntersectingEdge(polyLine.get(0), line))
                 {
-                    edges = splitRingLineIntoEdges(changeSet, line);
+                    // we just want a single Edge for the whole loop, connecting back to itself
```

Comment:
sort of. the refactor consolidates a lot of logic into one place so that we don't need an entirely different processing method for rings. because of that and some cleanup, there's a bit more consistency with ring handling, but it's all overwhelmingly similar to the original method. for example, the way section integration test ends up outputting an entirely equivalent atlas, despite a significant number of rings.